### ISSUE TITLE
Depending on compiler-builtins creates more unresolved symbols.

### ISSUE THREAD

**Issue by turnage** [rust-lang-nursery/compiler-builtins#123]
I am building for arm-none-eabihf (no_std) with the following target file:

````
{
    "llvm-target": "arm-none-eabihf",
    "target-endian": "little",
    "target-pointer-width": "32",
    "os": "none",
    "env": "eabi",
    "vendor": "unknown",
    "arch": "arm",

    "data-layout": "e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64",
    "executables": true,
    "relocation-model": "static",
    "no-compiler-rt": true
}
````

I build with ````RUST_TARGET_PATH=`pwd` xargo build --target arm-none-eabihf````.

Before adding a dependency on compiler-builtins, I get the following unresolved symbols, found by ````arm-none-eabi-nm target/arm-none-eabihf/debug/libtyran.a | grep "U "````:

````
         U _ZN4core9panicking5panic17h86aded962ecadfccE
         U __aeabi_memclr4
         U __aeabi_memcpy
         U __aeabi_memcpy4
         U __aeabi_memset
         U __aeabi_uidiv
         U __aeabi_uidivmod
         U __aeabi_ul2d
         U __aeabi_ul2f
         U __aeabi_uldivmod
         U __aeabi_unwind_cpp_pr0
         U __aeabi_unwind_cpp_pr1
         U memcmp
         U __mulodi4
         U rust_begin_unwind
         U __sync_val_compare_and_swap_1
         U __sync_val_compare_and_swap_2
         U __sync_val_compare_and_swap_4
````

After adding compiler-builtins to my dependencies, the extern crate statement and corresponding feature in my lib.rs, I get the following undefined symbols found by the same command:

````
         U _ZN4core9panicking5panic17h86aded962ecadfccE
         U __aeabi_ddiv
         U __aeabi_dmul
         U __aeabi_fdiv
         U __aeabi_fmul
         U __aeabi_unwind_cpp_pr0
         U __aeabi_unwind_cpp_pr1
         U memcpy
         U memmove
         U memset
         U rust_eh_personality
         U _Unwind_Resume
         U _ZN4core9panicking5panic17h86aded962ecadfccE
         U __aeabi_memclr4
         U __aeabi_memcpy
         U __aeabi_memcpy4
         U __aeabi_memset
         U __aeabi_uidiv
         U __aeabi_uidivmod
         U __aeabi_ul2d
         U __aeabi_ul2f
         U __aeabi_uldivmod
         U __aeabi_unwind_cpp_pr0
         U __aeabi_unwind_cpp_pr1
         U memcmp
         U __mulodi4
         U rust_begin_unwind
         U __sync_val_compare_and_swap_1
         U __sync_val_compare_and_swap_2
         U __sync_val_compare_and_swap_4
````

Am I misusing this crate? What is the expected behavior in this case?

**Comment by japaric** [comment]
I don't use `staticlib`s much but I don't think it's going to give you what *I think* you want to produce: a standalone (archived) (flashable) object file because building staticlibs doesn't involve the linker at all and that's what takes cares of "filling" the undefined references you are seeing.

My understanding is that building staticlibs is just going to (re)archive together all the involved object files but not try to resolve references between them.

What does `arm-none-eabi-nm target/arm-none-eabihf/debug/libtyran.a | grep "T "` output? I expect you'll see some of the "undefined" symbol in there. e.g. `T __aeabi_uidiv` (note: that's a "defined" symbol).

**Comment by turnage** [comment]
I did not know that building static libs did not invoke the linker!

The offending code that was creating references to compiler-rt symbols was:

````
for i in 0..10 {}
````

When building an rlib, this code stopped creating references to compiler-rt symbols.

There were still the following undefined symbols:

````
undefined reference to `core::panicking::panic::h86aded962ecadfcc'
undefined reference to `__aeabi_unwind_cpp_pr0'
undefined reference to `_Unwind_Resume'
````

But these references were removed with lto.

Thank you for the response @japaric!
